src/common/services/chain-client.service.ts

```diff
@@ -4,36 +4,32 @@ import { ConfigService } from '@nestjs/config';
 import { Chain } from '@types';
 import { celo, mainnet } from 'viem/chains';
 import { Logger } from '@nestjs/common';
+import { Semaphore } from '@/utils';
 
 @Injectable()
 export class ChainClientService {
-  private clients: Map<Chain, PublicClient> = new Map();
-  private blockWatchers: Map<Chain, () => void> = new Map();
+  private clients = new Map<Chain, PublicClient>();
+  private blockWatchers = new Map<Chain, () => void>();
+  private rpcLimiters = new Map<Chain, Semaphore>();
+  private globalRpcLimiter = new Semaphore(1);
```

Comment:
it might be worth investigating whether we should just replace our own handrolled rate limiting logic with a battle-tested library, couple options from sindre:
- [p-limit](https://www.npmjs.com/package/p-limit) (minimal option just for limiting concurrency)
- [p-throttle](https://www.npmjs.com/package/p-throttle) for rate-limiting individual function calls
- [p-queue](https://www.npmjs.com/package/p-queue) full queue